### PR TITLE
add support for the createSecret restful API

### PR DESCRIPTION
**Pull request by Yidingluo1**
this restful API will be exposed to public as:
POST <ehsm_srv_address>/ehsm?Action=CreateSecret
add the corresponding test cases.
test: passed the python test

Signed-off-by: wanghouqi <wanghouqi@neusoft.com>

### CODE REVIEW

File: ehsm_kms_service/constant.js [hunk 1]
```diff
@@ -6,6 +6,8 @@ const CMK_EFFECTIVE_DURATION = 1 * 365 * 24 * 60 * 60 * 1000
 const CMK_LOOP_CLEAR_TIME = 24 * 60 * 60 * 1000
 const CMK_EXPIRE_TIME_EXPAND = 10 * 24 * 60 * 60 * 1000
 const CMK_LOOP_CLEAR_EXECUTION_TIME = 3
+const SM_SECRET_VERSION_STAGE_CURRENT = 'Current'
+const SM_SECRET_VERSION_STAGE_PREVIOUS = 'Previous'
```
[the remark this answers is not in the document]
Comment (wanghouqi) [reply ehsm_kms_service/constant.js:10]: no,'Current' and 'Previous' will be compared with hte parameters passed in by the user.

File: ehsm_kms_service/secret_manager_apis.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        let secret_name_result = await DB.partitionedFind('secret_metadata', secret_name_query)
+        if (secret_name_result.docs.length > 0) {
+            res.send(_result(400, 'Secret name under appid is unique'))
+            return
+        }
+
+        if (encryptionKeyId != '' && typeof (encryptionKeyId) == 'string') {
+            const query_key_id = {
+                selector: {
+                    keyid: encryptionKeyId,
```
[the remark this answers is not in the document]
Comment (wanghouqi) [reply ehsm_kms_service/secret_manager_apis.js:110]: appid will be add to selector

File: ehsm_kms_service/secret_manager_apis.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            selector: {
+                appid,
+                secretName
+            },
+            fields: ['appid', 'secretName'],
+            limit: 1,
+        }
+        let secret_name_result = await DB.partitionedFind('secret_metadata', secret_name_query)
+        if (secret_name_result.docs.length > 0) {
+            res.send(_result(400, 'Secret name under appid is unique'))
```
[the remark this answers is not in the document]
Comment (wanghouqi) [reply ehsm_kms_service/secret_manager_apis.js:103]: It will be change

File: ehsm_kms_service/secret_manager_apis.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            return
+        }
+        const apikey_encrypt_res = napi_result(cryptographic_apis.Encrypt, res, [keyBlob, secretData, ''])
+        const { ciphertext } = apikey_encrypt_res.result
+
+        DB.insert({
+            _id: `secret_metadata:${uuidv4()}`,
+            appid,
+            secretName,
+            encryptionKeyId,
```
[the remark this answers is not in the document]
Comment (wanghouqi) [reply ehsm_kms_service/secret_manager_apis.js:147]: We have to verify that the current incoming data type is string

[the remark this answers is not in the document]
Comment (HawieLong) [reply ehsm_kms_service/secret_manager_apis.js:147]: Let's check this data type not only use "string", such as convert to base64 first or set the network transfer type to base64. Make sure this cannot be used as a database insert attack.

File: ehsm_kms_service/secret_manager_apis.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    } else {
+        if (typeof (param) == 'string') {
+            return true
+        } else {
+            return false
+        }
+    }
+}
+
+//creator Secret
```
Comment (HawieLong) [inline ehsm_kms_service/secret_manager_apis.js:56]: create

File: ehsm_kms_service/secret_manager_apis.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (HawieLong) [reply ehsm_kms_service/secret_manager_apis.js:147]: And the secret name may need a base64 encoding.

File: ehsm_kms_service/secret_manager_apis.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                versionStage: SM_SECRET_VERSION_STAGE_CURRENT
+            }).then(() => {
+                if (tags_jsonArray.length > 0) {
+                    try {
+                        for (let i in tags_jsonArray) {
+                            DB.insert({
+                                _id: `secret_tags:${uuidv4()}`,
+                                appid,
+                                secretName,
+                                tagKey: tags_jsonArray[i]['tagKey'],
```
Comment (HawieLong) [inline ehsm_kms_service/secret_manager_apis.js:173]: tags need using base64 as well.

File: test/test_kms_with_cli.py [hunk 7]
```diff
@@ -178,4 +192,8 @@ def get_args():
     test_deleteAllKey(base_url, headers)
 
     test_GenerateQuote_and_VerifyQuote(base_url, headers)
+
+    test_secret_manager(base_url, headers)
```
[the remark this answers is not in the document]
Comment (wanghouqi) [reply test/test_kms_with_cli.py:196]: the  test_secret_manager will be move to new py file test_secret_with_cli.py

File: ehsm_kms_service/secret_manager_apis.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (wanghouqi) [reply ehsm_kms_service/secret_manager_apis.js:147]: maybe ,we can base64 all parameter data?

File: test/test_kms_with_cli.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (HawieLong) [reply test/test_kms_with_cli.py]: Good, a new file is better. The file name can be test_secret_manager_with_cli.py

File: ehsm_kms_service/secret_manager_apis.js [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+        DB.insert({
+            _id: `secret_metadata:${uuidv4()}`,
+            appid,
+            secretName,
+            encryptionKeyId,
+            description,
+            createTime,
+            deleteTime: null,
+            plannedDeleteTime: null,
+            rotationInterval,
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:236]: rotation support will be in another patch, right?

File: ehsm_kms_service/secret_manager_apis.js [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+        let rotationInterval = getParam_String(payload, 'rotationInterval')
+        if (!checkStringParam(secretData, true)) {
+            res.send(_result(400, 'secretData cannot be empty and must be string'))
+            return
+        }
+        if (!checkStringParam(secretName, true)) {
+            res.send(_result(400, 'secretName cannot be empty and must be string'))
+            return
+        }
+        if (!checkStringParam(versionId, true)) {
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:163]: versionId is a MUST string. 
if versionid is +_+, what's the next versionId after auto-rotation?

File: ehsm_kms_service/secret_manager_apis.js [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+function getParam_String(payload, key, needBase64) {
+    if (needBase64 == undefined) {
+        needBase64 = true;
+    }
+    let val = payload[key]
+    if (val != undefined) {
+        val = String(val)
+        if (needBase64 && val.length > 0) {
+            if (isJSONString(val)) {
+                val = toBase64(eval("(" + val + ")"))
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:51]: eval is dangerous, with high-security concerns. 
Please review the below and replace it with others, like Function().
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#never_use_eval!

File: ehsm_kms_service/secret_manager_apis.js [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+                json[key] = toBase64(element)
+            }
+        }
+        return json;
+    }
+}
+
+function isJSONString(val) {
+    try {
+        let json = eval("(" + val + ")")
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:84]: same

File: ehsm_kms_service/secret_manager_apis.js [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+} = require('./function')
+const {
+    cryptographic_apis,
+} = require('./apis')
+const ehsm_napi = require('./ehsm_napi')
+
+//check tags format
+function buildTagsArray(tags) {
+    try {
+        var tags_jsonArray = eval(tags)
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:21]: same

File: ehsm_kms_service/secret_manager_apis.js [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+        }
+        if (!checkStringParam(rotationInterval, false)) {
+            res.send(_result(400, 'rotationInterval must be string'))
+            return
+        }
+        if (!checkStringParam(tags, false)) {
+            res.send(_result(400, 'tags must be string'))
+            return
+        }
+        if (!((/^([a-z\d]{8}-[a-z\d]{4}-[a-z\d]{4}-[a-z\d]{4}-[a-z\d]{12})$/).test(encryptionKeyId))) {
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:179]: create a function e.g. check_uuid_format() or others.

File: ehsm_kms_service/secret_manager_apis.js (hunk 8 above)
Comment (wanghouqi) [reply ehsm_kms_service/secret_manager_apis.js:236]: Yes, now only one field is reserved

File: ehsm_kms_service/secret_manager_apis.js (hunk 10 above)
Comment (wanghouqi) [reply ehsm_kms_service/secret_manager_apis.js:51]: will be change to Function

File: ehsm_kms_service/secret_manager_apis.js [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+        }
+        if (!checkStringParam(description, false)) {
+            res.send(_result(400, 'description must be string'))
+            return
+        }
+        if (!checkStringParam(rotationInterval, false)) {
+            res.send(_result(400, 'rotationInterval must be string'))
+            return
+        }
+        if (!check_encryptionKeyId_format(encryptionKeyId) && encryptionKeyId != '' && encryptionKeyId != undefined) {
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:109]: seems the logic is wrong?

is the check required: encryptionKeyId != '', can be covered by regex?  or move to the function check_encryptionKeyId_format() ?

File: ehsm_kms_service/secret_manager_apis.js [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+                    nextRotationDate = timenum * 24 * 60 * 60 * 1000 + createTime
+                    break
+                case (unit = 'h'):
+                    nextRotationDate = timenum * 60 * 60 * 1000 + createTime
+                    break
+                case (unit = 'm'):
+                    nextRotationDate = timenum * 60 * 1000 + createTime
+                    break
+                case (unit = 's'):
+                    nextRotationDate = timenum * 1000 + createTime
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:167]: just one open: if set 1s, will result in DoS?

Comment (wanghouqi) [reply ehsm_kms_service/secret_manager_apis.js:167]: Yes, will be add a rule, the rotationInterval must between 6h and 8760h

**Review by yang8621 — APPROVED** [review]

**Comment by wanghouqi** [comment]
> see, encryptionKeyId is optional, rig

Yes

File: ehsm_kms_service/secret_manager_apis.js [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+            } else {
+                keyBlob = key_id_res.docs[0].keyBlob
+            }
+        } else if (encryptionKeyId == "" || encryptionKeyId == undefined) {
+            let default_cmk_res = await DB.partitionedFind('user_info', queryDefaultCMK(appid))
+            if (default_cmk_res.docs.length == 0) {
+                res.send(_result(400, 'Cannot find default CMK'))
+                return
+            }
+            keyBlob = default_cmk_res.docs[0].sm_default_cmk
```
Comment (syan10) [inline ehsm_kms_service/secret_manager_apis.js:150]: so the sm_default_cmk is a keyblob rather than a cmk_id?

File: ehsm_kms_service/secret_manager_apis.js [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+        } else {
+            res.send(_result(400, 'encryptionKeyId error'))
```
Comment (syan10) [inline ehsm_kms_service/secret_manager_apis.js:152]: should never run here, it's a internal error, right? since the default cmk should always exist.

File: ehsm_kms_service/secret_manager_apis.js [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+                    break
+                case (unit = 's'):
+                    nextRotationDate = timenum * 1000 + createTime
+                    break
+            }
+            let interval = nextRotationDate - createTime
+
+            console.info('interval', interval)
+            if (6 * 60 * 60 * 1000 > interval || interval > 8760 * 60 * 60 * 1000) {
+                res.send(_result(400, 'rotationInterval is too short'))
```
Comment (syan10) [inline ehsm_kms_service/secret_manager_apis.js:178]: the error msg is not make sense.

File: ehsm_kms_service/secret_manager_apis.js [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+                res.send(_result(400, 'Cannot find default CMK'))
+                return
+            }
+            keyBlob = default_cmk_res.docs[0].sm_default_cmk
+        } else {
+            res.send(_result(400, 'encryptionKeyId error'))
+            return
+        }
+        const apikey_encrypt_res = napi_result(cryptographic_apis.Encrypt, res, [keyBlob, secretData, ''])
+        const { ciphertext } = apikey_encrypt_res.result
```
Comment (syan10) [inline ehsm_kms_service/secret_manager_apis.js:156]: suggest to put the rotationInterval codes into a sperate function.

File: test/test_secret_manager_with_cli.py [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+    appid, apikey = enroll.enroll(base_url)
+    _utils_.init_appid_apikey(appid, apikey)
+
+def test_secret_manager(base_url, headers):
+    
+    print('====================test_secret_manager start===========================')
+    key1 = createkey.createkey(base_url, "EH_AES_GCM_128", "EH_INTERNAL_KEY")
+
+    createSecret.createSecret(base_url, "secretData1", "secret001", key1, "mysecret", "30h")
+
```
Comment (syan10) [inline test/test_secret_manager_with_cli.py:30]: you should add another case without provide the cmk.

File: ehsm_kms_service/secret_manager_apis.js (hunk 16 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:150]: sm_ default_ CMK is used for encryption as keyblob here

File: ehsm_kms_service/secret_manager_apis.js (hunk 17 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:152]: else{} will be deleted

File: ehsm_kms_service/secret_manager_apis.js (hunk 18 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:178]: it will change to 'The rotation time is less than 6 hours or more than 8760 hours'

File: test/test_secret_manager_with_cli.py (hunk 20 above)
Comment (Yidingluo1) [reply test/test_secret_manager_with_cli.py:30]: it will be add in test

File: ehsm_kms_service/secret_manager_apis.js (hunk 17 above)
Comment (syan10) [reply ehsm_kms_service/secret_manager_apis.js:152]: you don't need to delete the else{}, just update the error log, such as "internal error, should never run here..."

Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:152]: ok

File: ehsm_kms_service/secret_manager_apis.js [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+        if (typeof (param) == 'string') {
+            return true
+        } else {
+            return false
+        }
+    }
+}
+
+function check_encryptionKeyId_format(encryptionKeyId) {
+    if (encryptionKeyId != '' && encryptionKeyId != undefined) {
```
Comment (HawieLong) [inline ehsm_kms_service/secret_manager_apis.js:48]: if encryptionKeyId is null, will this return false?

Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:48]: Null will not appear, and it will be converted to string on line 22

File: test/cli/createSecret.py [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+    print(base_url)
+    return base_url, args.secretData, args.secretName, args.encryptionKeyId, args.description, args.rotationInterval
+
+def createSecret(base_url, secretData, secretName, encryptionKeyId, description, rotationInterval):
+    print('create secret')
+    
+    payload = OrderedDict()
+    payload["description"] = description
+    if encryptionKeyId != None:
+        payload["encryptionKeyId"] = encryptionKeyId
```
Comment (syan10) [inline test/cli/createSecret.py:35]: so how about the comman line looks like?

does the user need to provide the encryptionkeyid as None? can you show me the details about the cmdline?

**Review by HawieLong — APPROVED** [review]

**Review by syan10 — APPROVED** [review]
